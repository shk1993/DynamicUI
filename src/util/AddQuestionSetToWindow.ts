import QuestionSet from '../util/SampleQuestionSet';

declare global {
  // tslint:disable-next-line:interface-name
  interface Window {
    QuestionSet: typeof QuestionSet;
    ShowQuestionSet: () => void;
  }
}

/**
 * Appends the entire QuestionSet JSON to document.body
 */
const showQuestionSet = () => {
  const pre = document.createElement('pre');
  pre.setAttribute('id', 'questionset');
  pre.innerHTML = JSON.stringify(QuestionSet, undefined, 2);
  document.body.appendChild(pre);
};

/**
 * Exports the additional components to the global window variable
 */
export default () => {
  window.QuestionSet = window.QuestionSet || QuestionSet;
  window.ShowQuestionSet = window.ShowQuestionSet || showQuestionSet;
};
