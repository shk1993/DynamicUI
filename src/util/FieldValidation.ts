import { IAttribute } from '../interfaces/attributes';
import { IValidator } from '../interfaces/validator';

/**
 * This function returns a function that reduces all validators.
 * @param validators   array of validator functions
 * @returns A function that reduces the result of array of validtors
 */
const composeValidators = (...validators: IValidator[]) => (
  value: string | undefined,
) =>
  validators.reduce((error, validator) => error || validator(value), undefined);

/**
 * This function filters validation key from the attributes.
 * It returns the regex validator function if validation attribute is present or returns undefined
 * @param attributes   Attributes on the render field
 * @returns       A function that validates field value matches the regex value or undefined
 */
const getRegexValidator = (attributes: IAttribute[]) => {
  const validationAttr = attributes.filter(
    (attr: IAttribute) => attr.key === 'validation',
  );
  if (
    validationAttr &&
    validationAttr.length !== 0 &&
    validationAttr[0].value
  ) {
    const regex = new RegExp(validationAttr[0].value);
    return (value: string | undefined) =>
      !value || regex.test(value) ? undefined : 'Invalid Input';
  }
  return undefined;
};

/**
 * This function filters required key from the attributes.
 * It returns the required validator function required attribute value is true or returns undefined
 * @param attributes   Attributes on the render field
 * @returns       A function that validates field value is required or undefined
 */
const getRequiredValidator = (attributes: IAttribute[]) => {
  const reqAttr = attributes.filter(
    (attr: IAttribute) => attr.key === 'required',
  );
  if (
    reqAttr &&
    reqAttr.length !== 0 &&
    reqAttr[0].value &&
    reqAttr[0].value === 'true'
  ) {
    return (value: string | undefined) => (value ? undefined : 'Required');
  }
  return undefined;
};

/**
 * This function gets the regex validator, required validator based on the attributes and returns a reduced combimned function.
 * @param attributes   Attributes on the render field
 * @returns       A function that reduces errors on all validators
 */
const composeAllValidators = (attributes: IAttribute[]) => {
  let validatorFunctions: IValidator[] = [];
  if (attributes) {
    const regexFunc = getRegexValidator(attributes);
    const reqFunc = getRequiredValidator(attributes);
    if (regexFunc) {
      validatorFunctions = validatorFunctions.concat([regexFunc]);
    }
    if (reqFunc) {
      validatorFunctions = validatorFunctions.concat([reqFunc]);
    }
  }
  return composeValidators(...validatorFunctions);
};

/**
 * This function is the function to get the validators on the field. It returns a function that reduces errors on all validators.
 * @param attributes   Attributes on the render field
 * @returns       A function that reduces errors on all validators
 */
export const getValidators = (attributes: IAttribute[] | undefined) => {
  return attributes ? composeAllValidators(attributes) : undefined;
};
