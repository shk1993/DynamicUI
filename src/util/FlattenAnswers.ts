import { IAnswers } from '../interfaces/answers';

/**
 * Retrieves an object of answers, maps over all the keys, and then returns an array of answer objects
 * @param answers An object of key-object pairs, where key is the question id
 * @returns       An array of answer objects
 */
export default (answers: IAnswers) => {
  return Object.keys(answers).map(ans => answers[ans]);
};
