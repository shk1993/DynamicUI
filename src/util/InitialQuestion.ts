/**
 * Returns the initial question to start the question set form
 * @returns An array of strings, currently ['DE30'], for the questions to start
 */
export default () => ['DE01'];
