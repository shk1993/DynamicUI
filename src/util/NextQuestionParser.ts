import { Grammar, Parser } from 'nearley';
import { IAnswers } from '../interfaces/answers';
import {
  ICondition,
  IConditional,
  IDEIDFunction,
  INextQuestion,
  IQuestionReturn,
  IQuestionValue,
} from '../interfaces/nextQuestion';
import { map } from '../state/stateHandler';

let grammar: Grammar;
let parser: Parser;

// tslint:disable-next-line:no-var-requires
grammar = Grammar.fromCompiled(require('../grammar/nextQuestion'));

const isConditionTrue = (condition: ICondition) => {
  const result = map[condition.condEval.state][condition.expr](
    condition.condEval.QID,
    condition.AID,
  );
  return result !== undefined ? result : true;
};

const evaluateNextQuestionFunction = (deid: IDEIDFunction) => {
  const nextQuestion = map.loop.peek(deid.state.QID);
  map.loop.pop(deid.state.QID);
  return nextQuestion;
};

const evaluateConditional = (conditional: IConditional): INextQuestion => {
  return conditional.condition.every(isConditionTrue)
    ? conditional.trueNextQuestion
    : conditional.falseNextQuestion;
};

const parseNextQuestion = (nextQuestion: string) => {
  parser = new Parser(grammar);
  parser.feed(nextQuestion);
  return parser.results[0];
};

const getNextQuestion = (nextQuestion: IQuestionValue): IQuestionReturn => {
  return typeof nextQuestion.QID === 'string'
    ? (nextQuestion as IQuestionReturn)
    : {
        modifier: nextQuestion.modifier,
        QID: evaluateNextQuestionFunction(nextQuestion.QID),
      };
};

/**
 * This function is the function called by the NextQuestion function. It returns an IQuestionReturn object
 * for the NextQuestion function to return.
 * @param value   Currently submited value for the current question
 * @param answers All the possible answer choices for the current question
 * @returns       An IQuestionReturn object NextQuestion to return back to QuestionSet
 */
export const getNextQuestionOfString = (value: string, answers: IAnswers) => {
  const nextQuestion =
    value in answers ? answers[value].nextQuestion : answers.noKey.nextQuestion;
  let nextQuestionParsed = parseNextQuestion(nextQuestion);
  while (nextQuestionParsed.type !== 'nextQuestion') {
    nextQuestionParsed = evaluateConditional(
      nextQuestionParsed.value as IConditional,
    );
  }
  if (nextQuestionParsed.value.length) {
    const next = [];
    for (const val of nextQuestionParsed.value) {
      next.push(getNextQuestion(val.DEID));
    }
    return next;
  } else {
    return [getNextQuestion(nextQuestionParsed.value as IQuestionReturn)];
  }
};
