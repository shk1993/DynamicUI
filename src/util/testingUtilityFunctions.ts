export const silenceConsoleErrors = () => {
  /* silence -- Warning: Stateless function components cannot be given refs. Attempts to access this ref will fail. */
  /* tslint:disable-next-line */
  console.error = () => {};
  /* tslint:disable-next-line */
  console.warn = () => {};
};
