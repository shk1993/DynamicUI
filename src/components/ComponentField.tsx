import Typography from '@material-ui/core/Typography';
import * as React from 'react';
import ComponentList from '../components/ComponentList';
import ICurrentQuestion from '../interfaces/currentQuestion';
import { IRenderComponent } from '../interfaces/parseRender';
import { setHistory } from '../state/historyState';
import FlattenAnswers from '../util/FlattenAnswers';
import ParseRender from '../util/ParseRender';
import ReturnAnswers from '../util/ReturnAnswers';
import QuestionSet from '../util/SampleQuestionSet';

/**
 * The starting point of the main purpose of our application and called from the QuestionSet page.
 * @param props The current question, history values, and the push function
 * @returns     React component consisting of the question(s) and answer(s)
 */
export default (props: ICurrentQuestion) => {
  const renderGroup = [];
  setHistory(props.values);
  for (const question of props.currentQuestion) {
    const curQuestion = QuestionSet[question];
    const render = ParseRender(curQuestion.render);
    const flatAnswers = ReturnAnswers(
      render,
      FlattenAnswers(curQuestion.answers),
    );
    renderGroup.push({ question: curQuestion, answers: flatAnswers, render });
  }
  return (
    <div>
      {renderGroup.map(groupComp => {
        return (
          <div key={JSON.stringify(groupComp)}>
            <Typography variant="subheading">
              {groupComp.question.text}
            </Typography>
            {groupComp.render.map((comp: IRenderComponent) => {
              const Component = ComponentList[comp.component];
              return (
                <Component
                  key={JSON.stringify(comp)}
                  label={groupComp.question.text}
                  name={groupComp.question.key}
                  answers={groupComp.answers}
                  values={props.values}
                  attributes={comp.attributes}
                  push={props.push}
                />
              );
            })}
            <br />
          </div>
        );
      })}
    </div>
  );
};
