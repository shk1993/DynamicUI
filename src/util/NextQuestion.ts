import { Set, Stack } from 'immutable';
import { IQuestionReturn } from '../interfaces/nextQuestion';
import { IQuestionData } from '../interfaces/questions';
import { map } from '../state/stateHandler';
import { getNextQuestionOfString } from './NextQuestionParser';
import QuestionSet from './SampleQuestionSet';

/**
 * Returns an array of next question ids.
 * @param nextQuestions array of next question id and modifiers
 */
const getNextQuestions = (nextQuestions: IQuestionReturn[]) => {
  const nextQuestionIds = [];
  for (const question of nextQuestions) {
    nextQuestionIds.push(question.DEID);
  }
  return nextQuestionIds;
};

/**
 * The function that is used to return the next questions for the user to answer.
 * @param currentQuestions  An array of questions the user just previously answered
 * @param values            The historical state of all previous answers made by the user
 */
export default (
  currentQuestions: string[],
  values: IQuestionData,
): string[] => {
  const question = currentQuestions[currentQuestions.length - 1];
  const valueSubmitted = values[question];
  const answers = QuestionSet[question].answers;

  // condition #1: the value submitted is a string. The string is the answer key
  if (typeof valueSubmitted === 'string') {
    return getNextQuestions(getNextQuestionOfString(valueSubmitted, answers));
  }

  // condition #2: the value submitted is an array of 0 elements. We should use the 'noKey' answer key
  if (valueSubmitted.length === 0) {
    return getNextQuestions(getNextQuestionOfString('noKey', answers));
  }

  // condition #3: the value submitted is an array and we need to evaluate all next question ids
  let answerList = Set<string | string[]>();
  for (const answer of valueSubmitted) {
    const nextQuestion = getNextQuestionOfString(answer, answers);
    if (nextQuestion[0].modifier === null) {
      return getNextQuestions(nextQuestion);
    }
    answerList = answerList.add(getNextQuestions(nextQuestion));
  }
  const questionsToVisit = answerList.toArray();
  if (questionsToVisit.includes('review')) {
    questionsToVisit.splice(questionsToVisit.indexOf('review'), 1);
    questionsToVisit.push('review');
  }
  if (questionsToVisit.length > 0) {
    map.loop.push(question, Stack<string>(questionsToVisit));
    const nextQuestion = map.loop.peek(question);
    map.loop.pop(question);
    return [nextQuestion];
  }

  throw Error('No next question to navigate to...');
};
