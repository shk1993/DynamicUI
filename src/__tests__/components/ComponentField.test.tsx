import * as React from 'react';
import { render } from 'react-testing-library';
import ComponentField from '../../components/ComponentField';
import { silenceConsoleErrors } from '../../util/testingUtilityFunctions';

Date.now = jest.fn(() => new Date('2017-01-01T00:00:00+0000'));

beforeAll(() => silenceConsoleErrors());

describe('ComponentField component', () => {
  it('should render the datepicker on question DE30', () => {
    const { container } = render(
      <ComponentField currentQuestion={['DE30']} values={{}} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the radio buttons on question DE3', () => {
    const { container } = render(
      <ComponentField currentQuestion={['DE3']} values={{}} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the date and radio buttons on question DE9', () => {
    const { container } = render(
      <ComponentField currentQuestion={['DE9']} values={{}} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
