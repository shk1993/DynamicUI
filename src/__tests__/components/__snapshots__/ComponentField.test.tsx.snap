// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentField component should render the date and radio buttons on question DE9 1`] = `
<div>
  <div>
    <h3
      class="MuiTypography-root-1 MuiTypography-subheading-8"
    >
      What is the event discovery date/time?
    </h3>
    <div>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <div
            answers="[object Object]"
            attributes="[object Object],[object Object],[object Object]"
            class="MuiFormControl-root-25 MuiFormControl-fullWidth-28"
            input="[object Object]"
            meta="[object Object]"
            values="[object Object]"
          >
            <div
              class="MuiInput-root-29 MuiInput-fullWidth-36 MuiInput-formControl-30 MuiInput-underline-33"
            >
              <input
                aria-invalid="false"
                aria-required="false"
                class="MuiInput-input-37"
                name="DE9"
                placeholder="MM/DD/YYYY"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <label>
        <div>
          <span
            class="MuiButtonBase-root-57 MuiIconButton-root-51 MuiSwitchBase-root-47 MuiRadio-root-42 MuiRadio-colorSecondary-46"
            input="[object Object]"
            label="Unknown"
            meta="[object Object]"
            values="[object Object]"
          >
            <span
              class="MuiIconButton-label-56"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-60"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <input
                class="MuiSwitchBase-input-50"
                name="DE9"
                type="radio"
                value="UNK"
              />
            </span>
            <span
              class="MuiTouchRipple-root-66"
            />
          </span>
          

          Unknown
          <br />
        </div>
      </label>
    </div>
    <br />
  </div>
</div>
`;

exports[`ComponentField component should render the datepicker on question DE30 1`] = `
<div>
  <div>
    <h3
      class="MuiTypography-root-1 MuiTypography-subheading-8"
    >
      What is the date the Common Formats patient safety module was completed?
    </h3>
    <div>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <div
            answers="[object Object]"
            attributes="[object Object]"
            class="MuiFormControl-root-25 MuiFormControl-fullWidth-28"
            input="[object Object]"
            meta="[object Object]"
            values="[object Object]"
          >
            <div
              class="MuiInput-root-29 MuiInput-fullWidth-36 MuiInput-formControl-30 MuiInput-underline-33"
            >
              <input
                aria-invalid="false"
                aria-required="false"
                class="MuiInput-input-37"
                name="DE30"
                placeholder="MM/DD/YYYY"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <br />
  </div>
</div>
`;

exports[`ComponentField component should render the radio buttons on question DE3 1`] = `
<div>
  <div>
    <h3
      class="MuiTypography-root-1 MuiTypography-subheading-8"
    >
      What is being reported?
    </h3>
    <div>
      <label>
        <div>
          <span
            class="MuiButtonBase-root-57 MuiIconButton-root-51 MuiSwitchBase-root-47 MuiRadio-root-42 MuiRadio-colorSecondary-46"
            input="[object Object]"
            label="Incident: A patient safety event that reached the patient, whether or not the patient was harmed."
            meta="[object Object]"
            values="[object Object]"
          >
            <span
              class="MuiIconButton-label-56"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-60"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <input
                class="MuiSwitchBase-input-50"
                name="DE3"
                type="radio"
                value="A3"
              />
            </span>
            <span
              class="MuiTouchRipple-root-66"
            />
          </span>
          

          Incident: A patient safety event that reached the patient, whether or not the patient was harmed.
          <br />
        </div>
      </label>
      <label>
        <div>
          <span
            class="MuiButtonBase-root-57 MuiIconButton-root-51 MuiSwitchBase-root-47 MuiRadio-root-42 MuiRadio-colorSecondary-46"
            input="[object Object]"
            label="Near Miss: A patient safety event that did not reach the patient."
            meta="[object Object]"
            values="[object Object]"
          >
            <span
              class="MuiIconButton-label-56"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-60"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <input
                class="MuiSwitchBase-input-50"
                name="DE3"
                type="radio"
                value="A6"
              />
            </span>
            <span
              class="MuiTouchRipple-root-66"
            />
          </span>
          

          Near Miss: A patient safety event that did not reach the patient.
          <br />
        </div>
      </label>
      <label>
        <div>
          <span
            class="MuiButtonBase-root-57 MuiIconButton-root-51 MuiSwitchBase-root-47 MuiRadio-root-42 MuiRadio-colorSecondary-46"
            input="[object Object]"
            label="Unsafe Condition: Any circumstance that increases the probability of a patient safety event."
            meta="[object Object]"
            values="[object Object]"
          >
            <span
              class="MuiIconButton-label-56"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-60"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <input
                class="MuiSwitchBase-input-50"
                name="DE3"
                type="radio"
                value="A9"
              />
            </span>
            <span
              class="MuiTouchRipple-root-66"
            />
          </span>
          

          Unsafe Condition: Any circumstance that increases the probability of a patient safety event.
          <br />
        </div>
      </label>
    </div>
    <br />
  </div>
</div>
`;
